editor/components/inserter/menu.js

```diff
@@ -1,46 +1,100 @@
 /**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+import uuid from 'uuid/v4';
+
+/**
  * Internal dependencies
  */
 import './style.scss';
 import Dashicon from 'components/dashicon';
 
-function InserterMenu( { position = 'top' } ) {
-	const blocks = wp.blocks.getBlocks();
-	const blocksByCategory = blocks.reduce( ( groups, block ) => {
-		if ( ! groups[ block.category ] ) {
-			groups[ block.category ] = [];
-		}
-		groups[ block.category ].push( block );
-		return groups;
-	}, {} );
-	const categories = wp.blocks.getCategories();
+class InserterMenu extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.state = {
+			filterValue: ''
+		};
+		this.filter = this.filter.bind( this );
+	}
+
+	filter( event ) {
+		this.setState( {
+			filterValue: event.target.value
+		} );
+	}
 
-	return (
-		<div className={ `editor-inserter__menu is-${ position }` }>
-			<div className="editor-inserter__arrow" />
-			<div className="editor-inserter__content">
-				{ categories
-					.map( ( category ) => !! blocksByCategory[ category.slug ] && (
-						<div key={ category.slug }>
-							<div className="editor-inserter__separator">{ category.title }</div>
-							<div className="editor-inserter__category-blocks">
-								{ blocksByCategory[ category.slug ].map( ( { slug, title, icon } ) => (
-									<div key={ slug } className="editor-inserter__block">
-										<Dashicon icon={ icon } />
-										{ title }
-									</div>
-								) ) }
+	selectBlock( slug ) {
+		return () => {
+			this.props.onInsertBlock( slug );
+			this.props.onSelect();
+		};
+	}
+
+	render() {
+		const { position = 'top' } = this.props;
+		const blocks = wp.blocks.getBlocks();
+		const isShownBlock = block => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
```

Comment:
Another note is we should probably extract some of these utilities as standalone functions and unit test them. Fine for another pull request.